src/edu/stanford/nlp/tagger/maxent/TestSentence.java

```diff
@@ -357,15 +357,13 @@ private void setHistory(int current, History h, int[] tags) {
 
     int left = leftWindow();
     int right = rightWindow();
+    final TTags ttags = maxentTagger.tags;
 
-    for (int j = current - left; j <= current + right; j++) {
-      if (j < left) {
-        continue;
-      } //but shouldn't happen
-      if (j >= size + left) {
-        break;
-      } //but shouldn't happen
-      h.setTag(j - left, maxentTagger.tags.getTag(tags[j]));
+    // Loop range, with guards:
+    final int s = Math.max(current - left, left);
+    final int e = Math.min(current + right, size + left - 1);
+    for (int j = s; j <= e; j++) {
```

Comment:
Random question from a distant memory.  I recall that in C++, ++i and i++ have slightly different performance because the return value of i++ is different from the final value of i.  Is that true in Java?  I would expect ++j to be a tiny bit faster.

Reply:
Indeed ++i is supposed to be faster, unless the compiler manages to optimize this. Here I would bet that Java can optimize this away - and so should any decent C/C++ compiler.
In C++ it may make a difference if operator++ is a pretty complex function - but in Java, this only exists for primitives, hence it should not matter.
According to the answers to this post:
https://stackoverflow.com/questions/12396740/is-x-more-efficient-than-x-in-java
both will both produce an iinc bytecode; just potentially in a different order. Hence, it should produce the exact same byte code here and then cannot make a difference.

Reply:
Neat, thanks!